Report 403 and 404 if service worker failed to install.

itrush opened the pull request:
- Added error handling for service worker registration call;
- Performing an additional request to determine if there is an issue with service worker file configuration;
- Reporting an error to the api in case of 403 and 404 errors
   - on first page view only;
   - not from inside a popup;
- Added tests for corresponding pieces in ServiceWorkerManager and SubscriptionManager;
- Added the workaround to correct the type checking for custom errors derived from Error class (TypeScript specific).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/368)
<!-- Reviewable:end -->

jkasten2 commented:
Reviewed 24 of 26 files at r1.
Review status: 24 of 26 files reviewed, 2 unresolved discussions (waiting on @itrush and @jkasten2)

---

*[src/managers/ServiceWorkerManager.ts, line 20 at r1](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/368#-LEkZBtS6_m94afvJmg3:-LEkZBtS6_m94afvJmg4:bvfriec) ([raw file](https://github.com/onesignal/onesignal-website-sdk/blob/a9ef293f596dd8a9f372f841b417bdb2b7fd1ca9/src/managers/ServiceWorkerManager.ts#L20)):*
> ```XML
> import ProxyFrame from '../modules/frames/ProxyFrame';
> import ServiceWorkerRegistrationError from "../errors/ServiceWorkerRegistrationError"
> import fetch from "node-fetch";
> ```

Should node-fetch be in the SDK source? When running in the browser will this be redefining fetch?

---

*[src/managers/SubscriptionManager.ts, line 156 at r1](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/368#-LEkR4htCP4jFHPcZZ_g:-LEkR4htCP4jFHPcZZ_h:blhjvce) ([raw file](https://github.com/onesignal/onesignal-website-sdk/blob/a9ef293f596dd8a9f372f841b417bdb2b7fd1ca9/src/managers/SubscriptionManager.ts#L156)):*
> ```XML
>     deviceRecord.appId = this.config.appId;
> 
>     // TODO why override? It's already part of previous call.
> ```

Address comment

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/368)*
<!-- Sent from Reviewable.io -->

itrush commented:
Review status: 21 of 25 files reviewed, 2 unresolved discussions (waiting on @jkasten2 and @itrush)

---

*[src/managers/ServiceWorkerManager.ts, line 20 at r1](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/368#-LEkZBtS6_m94afvJmg3:-LEkf54j8mq_ufEK5Bds:bunkt2a) ([raw file](https://github.com/onesignal/onesignal-website-sdk/blob/a9ef293f596dd8a9f372f841b417bdb2b7fd1ca9/src/managers/ServiceWorkerManager.ts#L20)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Should node-fetch be in the SDK source? When running in the browser will this be redefining fetch?
</blockquote></details>

Removed. Not needed after all.

---

*[src/managers/SubscriptionManager.ts, line 156 at r1](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/368#-LEkR4htCP4jFHPcZZ_g:-LEkf8s88n_z9da7I85j:bss4a2b) ([raw file](https://github.com/onesignal/onesignal-website-sdk/blob/a9ef293f596dd8a9f372f841b417bdb2b7fd1ca9/src/managers/SubscriptionManager.ts#L156)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Address comment
</blockquote></details>

Removed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/368)*
<!-- Sent from Reviewable.io -->

jkasten2 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 5 files at r2.
Review status: all files reviewed, 2 unresolved discussions (waiting on @jkasten2)

---



*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/368#-:-LElRCh66n8zvBnZJxok:bnfp4nl)*
<!-- Sent from Reviewable.io -->
